lib/Dialect/Torch/Transforms/DecomposeComplexOps.cpp

```diff
@@ -563,6 +563,51 @@ class DecomposeAtenEmptyLikeOp : public OpRewritePattern<AtenEmptyLikeOp> {
 };
 } // namespace
 
+// Helper function for generating TorchConstantOp either float or int based
+// on `isFloatDtype` with constant value `val`.
+static Value createConstantOp(Location loc, PatternRewriter &rewriter,
+                              bool isFloatDtype, int64_t val) {
+  if (isFloatDtype)
+    return rewriter.create<Torch::ConstantFloatOp>(
+        loc, rewriter.getF64FloatAttr((double)val));
+  else
+    return rewriter.create<Torch::ConstantIntOp>(
+        loc, rewriter.getI64IntegerAttr(val));
+}
+
+namespace {
+// Both the `aten.arange` and `aten.arange.start` op are converted into
+// `aten.arange.start_step` op.
+template <typename OpTy>
+class DecomposeAtenArangeLikeOp : public OpRewritePattern<OpTy> {
+  using OpRewritePattern<OpTy>::OpRewritePattern;
+  LogicalResult matchAndRewrite(OpTy op,
+                                PatternRewriter &rewriter) const override {
+    Location loc = op.getLoc();
+    Value start, step;
+    Value end = op.end();
+    Value result = op.result();
+    BaseTensorType resultType = result.getType().cast<BaseTensorType>();
+    bool isFloatDtype =
+        resultType.hasDtype() && resultType.getDtype().isa<mlir::FloatType>();
+    // If the op is a AtenArangeStartOp then it has a start value. Otherwise we
+    // set it as default value 0.
+    if (isa<AtenArangeStartOp>(op)) {
```

Comment:
I'm always a bit unsure about combining templates and dynamic casts. The main advantage of a template is that it allows compile time polymorphism (which can have a speed benefit), but the dynamic cast counteracts the benefit by requiring type information during runtime. 

One way of avoiding this issue is to have the template be a helper function that takes as arguments the `op` and the `start` value, then the two classes for decomposing `aten.arange` and `aten.arange.start` can call this function. In other words, we are abstracting too early. What the two ops have in common is not the decomposition, but the decomposition once the `start` value has been determined